File: cpp/test/phonenumbers/matcher_test.cc
// Copyright (C) 2017 The Libphonenumber Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Tests that all implementations of MatcherApi are consistent.

#include "phonenumbers/matcher_api.h"

#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "phonenumbers/regex_based_matcher.h"
#include "phonenumbers/phonemetadata.pb.h"

namespace i18n {
namespace phonenumbers {

namespace {

string ToString(const PhoneNumberDesc& desc) {
  string str = "pattern: ";
  if (desc.has_national_number_pattern()) {
    str += desc.national_number_pattern();
  } else {
    str += "none";
  }
  return str;
}

void ExpectMatched(
    const MatcherApi& matcher,
    const string& number,
    const PhoneNumberDesc& desc) {
  EXPECT_TRUE(matcher.MatchNationalNumber(number, desc, false))
      << number << " should have matched " << ToString(desc);
  EXPECT_TRUE(matcher.MatchNationalNumber(number, desc, true))
      << number << " should have matched " << ToString(desc);
}

void ExpectInvalid(
    const MatcherApi& matcher,
    const string& number,
    const PhoneNumberDesc& desc) {
  EXPECT_FALSE(matcher.MatchNationalNumber(number, desc, false))
      << number << " should not have matched " << ToString(desc);
  EXPECT_FALSE(matcher.MatchNationalNumber(number, desc, true))
      << number << " should not have matched " << ToString(desc);
}

void ExpectTooLong(
    const MatcherApi& matcher,
    const string& number,
    const PhoneNumberDesc& desc) {
  EXPECT_FALSE(matcher.MatchNationalNumber(number, desc, false))
      << number << " should have been too long for " << ToString(desc);
  EXPECT_TRUE(matcher.MatchNationalNumber(number, desc, true))
      << number << " should have been too long for " << ToString(desc);
}

}  // namespace

class MatcherTest : public testing::Test {
 protected:
  void CheckMatcherBehavesAsExpected(const MatcherApi& matcher) const {
    PhoneNumberDesc desc;

    desc = CreateDesc("");
    // Test if there is no matcher data.
    ExpectInvalid(matcher, "1", desc);

    desc = CreateDesc("9\\d{2}");
    ExpectInvalid(matcher, "91", desc);
    ExpectInvalid(matcher, "81", desc);
    ExpectMatched(matcher, "911", desc);
    ExpectInvalid(matcher, "811", desc);
    ExpectTooLong(matcher, "9111", desc);
    ExpectInvalid(matcher, "8111", desc);

    desc = CreateDesc("\\d{1,2}");
    ExpectMatched(matcher, "2", desc);
    ExpectMatched(matcher, "20", desc);

    desc = CreateDesc("20?");
    ExpectMatched(matcher, "2", desc);
    ExpectMatched(matcher, "20", desc);

    desc = CreateDesc("2|20");
    ExpectMatched(matcher, "2", desc);
    // Subtle case where lookingAt() and matches() result in different end()s.
    ExpectMatched(matcher, "20", desc);
  }

 private:
  // Helper method to set national number fields in the PhoneNumberDesc proto.
  // Empty fields won't be set.
  PhoneNumberDesc CreateDesc(
      const string& national_number_pattern) const {
    PhoneNumberDesc desc;
    if (!national_number_pattern.empty()) {
      desc.set_national_number_pattern(national_number_pattern);
    }
    return desc;
  }
};

TEST_F(MatcherTest, RegexBasedMatcher) {
  RegexBasedMatcher matcher;
  CheckMatcherBehavesAsExpected(matcher);
}

}  // namespace phonenumbers
}  // namespace i18n

File: java/libphonenumber/test/com/google/i18n/phonenumbers/internal/MatcherTest.java
/*
 * Copyright (C) 2017 The Libphonenumber Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.i18n.phonenumbers.internal;

import com.google.i18n.phonenumbers.Phonemetadata.PhoneNumberDesc;

import junit.framework.TestCase;

/**
 * Tests that all implementations of {@code MatcherApi} are consistent.
 */
public class MatcherTest extends TestCase {
  public void testRegexBasedMatcher() {
    checkMatcherBehavesAsExpected(RegexBasedMatcher.create());
  }

  private void checkMatcherBehavesAsExpected(MatcherApi matcher) {
    PhoneNumberDesc desc = createDesc("");
    // Test if there is no matcher data.
    assertInvalid(matcher, "1", desc);

    desc = createDesc("9\\d{2}");
    assertInvalid(matcher, "91", desc);
    assertInvalid(matcher, "81", desc);
    assertMatched(matcher, "911", desc);
    assertInvalid(matcher, "811", desc);
    assertTooLong(matcher, "9111", desc);
    assertInvalid(matcher, "8111", desc);

    desc = createDesc("\\d{1,2}");
    assertMatched(matcher, "2", desc);
    assertMatched(matcher, "20", desc);

    desc = createDesc("20?");
    assertMatched(matcher, "2", desc);
    assertMatched(matcher, "20", desc);

    desc = createDesc("2|20");
    assertMatched(matcher, "2", desc);
    // Subtle case where lookingAt() and matches() result in different end()s.
    assertMatched(matcher, "20", desc);
  }

  // Helper method to set national number fields in the PhoneNumberDesc proto. Empty fields won't be
  // set.
  private PhoneNumberDesc createDesc(String nationalNumberPattern) {
    PhoneNumberDesc.Builder desc = PhoneNumberDesc.newBuilder();
    if (nationalNumberPattern.length() > 0) {
      desc.setNationalNumberPattern(nationalNumberPattern);
    }
    return desc.build();
  }

  private void assertMatched(MatcherApi matcher, String number, PhoneNumberDesc desc) {
    assertTrue(String.format("%s should have matched %s.", number, toString(desc)),
        matcher.matchNationalNumber(number, desc, false));
    assertTrue(String.format("%s should have matched %s.", number, toString(desc)),
        matcher.matchNationalNumber(number, desc, true));
  }

  private void assertInvalid(MatcherApi matcher, String number, PhoneNumberDesc desc) {
    assertFalse(String.format("%s should not have matched %s.", number, toString(desc)),
        matcher.matchNationalNumber(number, desc, false));
    assertFalse(String.format("%s should not have matched %s.", number, toString(desc)),
        matcher.matchNationalNumber(number, desc, true));
  }

  private void assertTooLong(MatcherApi matcher, String number, PhoneNumberDesc desc) {
    assertFalse(String.format("%s should have been too long for %s.", number, toString(desc)),
        matcher.matchNationalNumber(number, desc, false));
    assertTrue(String.format("%s should have been too long for %s.", number, toString(desc)),
        matcher.matchNationalNumber(number, desc, true));
  }

  private String toString(PhoneNumberDesc desc) {
    StringBuilder strBuilder = new StringBuilder("pattern: ");
    if (desc.hasNationalNumberPattern()) {
      strBuilder.append(desc.getNationalNumberPattern());
    } else {
      strBuilder.append("none");
    }
    return strBuilder.toString();
  }
}
